### PR TITLE
[CAS/UnifiedOnDiskCache] Introduce `UnifiedOnDiskCache`

### PR DESCRIPTION
**Pull request by akyrtzi**
`UnifiedOnDiskCache` is a unified CAS nodes and key-value database, using on-disk storage for both. It manages storage growth and provides APIs for garbage collection.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: llvm/lib/CAS/UnifiedOnDiskCache.cpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+#include "llvm/CAS/OnDiskKeyValueDB.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+
+using namespace llvm;
+using namespace llvm::cas;
+using namespace llvm::cas::ondisk;
+
+/// FIXME: When the version of \p DBDirPrefix is bumped up we need to figure out
```
Comment (cachemeifyoucan) [inline llvm/lib/CAS/UnifiedOnDiskCache.cpp:63]: I am not sure we need to worry about this.

File: llvm/include/llvm/CAS/UnifiedOnDiskCache.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+///
+/// Usage patterns should be that an instance of \p UnifiedOnDiskCache is open
+/// for a limited period of time, e.g. for the duration of a build operation.
+/// For long-living processes that need periodic access to a
+/// \p UnifiedOnDiskCache, the client should device a scheme where access is
+/// performed within some defined period. For example, if a service is designed
+/// to continuously wait for requests that access a \p UnifiedOnDiskCache, it
+/// could keep the instance alive while new requests are coming in but close it
+/// after a time period in which there are no new requests.
+class UnifiedOnDiskCache {
```
Comment (cachemeifyoucan) [inline llvm/include/llvm/CAS/UnifiedOnDiskCache.h:39]: I would like to see this share the same abstraction with `RemoteCachingService`.

Maybe we can have `CachingService` base class with UnifiedOnDiskCache and RemoteCachingService as two different implementation?

**Review by cachemeifyoucan — COMMENTED** [review]
LGTM in general with some comments.

File: llvm/lib/CAS/UnifiedOnDiskCache.cpp (hunk 1 above)
Comment (akyrtzi) [reply llvm/lib/CAS/UnifiedOnDiskCache.cpp:63]: My concern is for the situation where a user gets a new toolchain and now their CAS/cache directory is permanently taking significant storage, until the user manually deletes it.
For example, let's say I set the limit to 100G, after getting a new toolchain (assuming the dir version is bumped) the directory now goes to 200G and garbage collection just makes it go back to 100G.

We could consider making directories of older versions eligible for deletion, but as the FIXME says, I'm punting this decision for later.

File: llvm/include/llvm/CAS/UnifiedOnDiskCache.h (hunk 2 above)
Comment (akyrtzi) [reply llvm/include/llvm/CAS/UnifiedOnDiskCache.h:39]: We don't need to worry about `RemoteCachingService`, my intention is to deprecate it in favor of a C API plugin interface, and eventually remove it.
Overall the client code in the compilers are only going to be interfacing with `ObjectStore` & `ActionCache`.

**Review by cachemeifyoucan — APPROVED** [review]

**Pull request merged**